interpreter/beam/beam.go

```diff
@@ -0,0 +1,510 @@
+		if pc < midStart {
+			highIdx = midIdx
+		} else if pc >= midEnd {
+			lowIdx = midIdx + 1
+		} else {
+			ertsCodeInfo = midStart
+			functionIndex = midIdx
+
+			// `mfa` is defined here:
+			// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/code_ix.h#L87-L95
+			moduleID := i.rm.Uint32(ertsCodeInfo + libpf.Address(16))
+			functionID := i.rm.Uint32(ertsCodeInfo + libpf.Address(16+8))
+			arity := i.rm.Uint32(ertsCodeInfo + libpf.Address(16+16))
+
+			return functionIndex, moduleID, functionID, arity, nil
+		}
+	}
+	lowStart := i.rm.Ptr(codeHeader + libpf.Address(136))
+	highEnd := i.rm.Ptr(codeHeader + libpf.Address(136+(highIdx+1)*8))
+	return 0, 0, 0, 0, fmt.Errorf("BEAM unable to find the MFA for PC 0x%x in expected code range (0x%x - 0x%x)", pc, lowStart, highEnd)
+}
+
+func (i *beamInstance) findFileLocation(codeHeader libpf.Address, functionIndex uint64, pc libpf.Address) (fileName string, lineNumber uint64, err error) {
+	lineTable := i.rm.Ptr(codeHeader + libpf.Address(72))
+
+	// `lineTable` points to a table of `BeamCodeLineTab_` structs:
+	// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_code.h#L130-L138
+	// (gdb) ptype/o $hdr.line_table
+	// type = const struct BeamCodeLineTab_ {
+	// 	     0      |       8     const Eterm *fname_ptr;
+	// 	     8      |       4     int loc_size;
+	// 	XXX  4-byte hole
+	// 	    16      |       8     union {
+	// 	                    8         Uint16 *p2;
+	// 	                    8         Uint32 *p4;
+	// 									   total size (bytes):    8
+	// 								   } loc_tab;
+	// 	    24      |       8    const void **func_tab[1];
+	//
+	// total size (bytes):   32
+
+	lineLow := i.rm.Ptr(lineTable + libpf.Address(8*functionIndex+24))
+	lineHigh := i.rm.Ptr(lineTable + libpf.Address(8*(functionIndex+1)+24))
+
+	// We need to align the lineMid values on 8-byte address boundaries
+	bitmask := libpf.Address(^(uint64(0xf)))
+	for lineHigh > lineLow {
+		lineMid := lineLow + ((lineHigh-lineLow)/2)&bitmask
+
+		if pc < i.rm.Ptr(lineMid) {
+			lineHigh = lineMid
+		} else if pc < i.rm.Ptr(lineMid+libpf.Address(8)) {
+			funcTab := i.rm.Ptr(lineTable + libpf.Address(24))
+			locIndex := uint32((lineMid - funcTab) / 8)
+			locSize := i.rm.Uint32(lineTable + libpf.Address(8))
+			locTab := i.rm.Ptr(lineTable + libpf.Address(16))
+			locAddr := locTab + libpf.Address(locSize*locIndex)
+			loc := uint64(0)
+			if locSize == 2 {
+				loc = uint64(i.rm.Uint16(locAddr))
+			} else {
+				loc = uint64(i.rm.Uint32(locAddr))
+			}
+			fnameIndex := loc >> 24
+			fileNamePtr := libpf.Address(i.rm.Uint64(lineTable) + 8*fnameIndex)
+			fileName = i.readErlangString(libpf.Address(i.rm.Uint64(fileNamePtr)), 256)
+
+			return fileName, loc & ((1 << 24) - 1), nil
+		} else {
+			lineLow = lineMid + 8
+		}
+	}
+
+	return "", 0, fmt.Errorf("BEAM unable to find file and line number")
+}
+
+func (i *beamInstance) lookupAtom(index uint32) (string, error) {
```

Comment:
This and possibly other similar "address to data" type of methods, might use LRU caching.

Reply:
Yep, thanks for the reminder. I was intending to figure that out but I'm a Golang noob so I didn't research how that's typically done yet.

Reply:
Various other interpreters do it, for one example you can look at  https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/cbb892e867e39715d4fefc4a945e6fd02e31c1a9/interpreter/nodev8/v8.go#L908